### PR TITLE
[tinker] Support PPO loss with Tinker and add critic model in SkyRLTrainBackend

### PR DESCRIPTION
**Pull request by tamoghnokandar**
Fixes the first issue of #1380.

## Summary                                                                                                            
                                                                                                                        
  - Add critic model support to the Tinker backend, enabling actor-critic PPO training through the Tinker API           
  - Extend `LossFnInputs` with optional `values` and `returns` fields, and add `ppo_critic` loss type                   
  - Refactor `SkyrlTrainBackend` from single-model to multi-model registry (`model_id → role`), supporting both policy  
  and critic actor groups
  - Critic models are created via `create_model` with `model_role="critic"`, sharing the policy's base model with       
  independent training                                                                                                  
  - Add `register_actor_group()` and `set_algorithm_config()` to `WorkerDispatch` for dynamic critic registration
                                                                                                                        
  ## Test plan 

- [x] Unit tests for new `LossFnInputs` fields (`values`, `returns`) and `Datum.to_types()` conversion                
- [x] Unit tests for `prepare_model_pass_batch` with `ppo_critic` loss type
- [x] Unit tests verifying JAX backend rejects critic role and `ppo_critic` loss                                      
- [x] Verify existing policy-only workflows remain unaffected (values/returns default to empty)         
<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1389" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

### CODE REVIEW

File: skyrl/backends/skyrl_train_backend.py [hunk 1]
```diff
@@ -182,6 +198,36 @@ def build_models(self, PolicyWorker):
...
+            num_gpus_per_actor=0.2 if colocate_all else 1,
+            colocate_all=colocate_all,
+            sequence_parallel_size=cfg.trainer.critic.sequence_parallel_size,
+        )
+        self._dispatch.register_actor_group("critic", critic_model)
+        self._dispatch.init_model("critic", cfg.trainer.critic.model.path, num_training_steps=1e9)
+        ray.get(critic_model.async_run_ray_method("pass_through", "_set_pad_token_id", self._tokenizer.pad_token_id))
+        if colocate_all:
+            critic_model.offload_to_cpu()
+            self._dispatch.mark_all_offloaded()
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train_backend.py]: instead of a `mark_all_offloaded` call here we should just have  a `mark_as_offloaded` method that marks a specific `model_id` as offloaded.

File: skyrl/backends/skyrl_train_backend.py [hunk 2]
```diff
@@ -114,18 +113,35 @@ def __init__(self, base_model: str, config: SkyRLTrainBackendOverrides):
         self.base_model = base_model
         # NOTE: We currently have two config attributes "config" which is just config overrides and "_cfg" which is the actual config object. This is a temporary state given that the Tinker engine expects a .config attribute
         self.config = config
-        self._model_id: str | None = None
-        self._model_metadata: types.ModelMetadata | None = None
+        self._model_ids: dict[str, str] = {}
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train_backend.py]: This should be renamed to `_model_ids_to_role`

File: skyrl/backends/skyrl_train_backend.py [hunk 3]
```diff
@@ -114,18 +113,35 @@ def __init__(self, base_model: str, config: SkyRLTrainBackendOverrides):
...
+            raise ValueError(f"Model {model_id} not found") from exc
+
+    def _get_batch_role(self, model_ids: list[str]) -> str:
+        if not model_ids:
+            return "policy"
+        roles = {self._get_role(model_id) for model_id in model_ids}
+        if len(roles) != 1:
+            raise ValueError(f"Mixed model roles in one batch are not supported: {sorted(roles)}")
+        if len(set(model_ids)) != 1:
+            raise ValueError(f"Mixed model_ids in one batch are not supported: {sorted(set(model_ids))}")
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train_backend.py]: Aren't these checks redundant ? If `len(set(model_ids) == 1`, then `len(roles)` will also be 1.

File: skyrl/backends/skyrl_train_backend.py [hunk 4]
```diff
@@ -458,33 +554,37 @@ def forward(
...
         results = {}
         for request_id, _, start_idx, end_idx in prepared_batch.request_batch_slices:
             loss_fn_outputs = []
             for i in range(start_idx, end_idx):
                 # Use token weights length to determine each example's actual response length
                 valid_len = len(prepared_batch.all_token_weights[i])
-                start = max(output_logprobs.shape[1] - valid_len, 0)
-                logprobs = output_logprobs[i, start:].tolist()
+                start = max(output_tensor.shape[1] - valid_len, 0)
+                values = output_tensor[i, start:].tolist()
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train_backend.py]: Nit: This should just be `outputs` ? can be values or logprobs

File: skyrl/backends/skyrl_train_backend.py [hunk 5]
```diff
@@ -528,9 +627,15 @@ def sample(
...
         # 2. Validate single model
         unique_models = set(prepared_batch.all_model_ids)
-        if unique_models != {self._model_id}:
+        if len(unique_models) != 1:
+            error = types.ErrorResponse(
+                error=f"Expected exactly one model_id for sampling, got {unique_models}", status="error"
+            )
+            return {req_id: error for req_id, _, _, _, _ in prepared_batch.request_batch_slices}
+        model_id = next(iter(unique_models))
+        if self._get_role(model_id) != "policy":
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/backends/skyrl_train_backend.py]: Good catch

File: skyrl/tinker/loss_fns.py [hunk 6]
```diff
@@ -78,12 +78,23 @@ def cispo_loss(
...
+def ppo_critic_loss(
+    _target_logprobs: jax.Array,
+    _loss_mask: jax.Array,
+    _sampling_logprobs: jax.Array,
+    _advantages: jax.Array,
+    _loss_fn_config: LossFnConfig,
+) -> jax.Array:
+    return jnp.zeros_like(_loss_mask)
+
+
```
Comment (SumanthRH) [inline skyrl/tinker/loss_fns.py]: This is not implemented?

Comment (tamoghnokandar) [reply skyrl/tinker/loss_fns.py:90]: Yes, this is not implemented. Should I add critic model support for JAX backend too?

File: skyrl/backends/skyrl_train/workers/worker_dispatch.py [hunk 7]
```diff
@@ -70,6 +70,10 @@ def __init__(
         # GPU state tracking (only matters when colocated)
         self._gpu_state: Dict[str, GPUState] = {name: GPUState() for name in self._actor_groups.keys()}
 
+    def register_actor_group(self, model: str, actor_group: PPORayActorGroup) -> None:
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/workers/worker_dispatch.py:73]: Nit: docstring

**Comment by tamoghnokandar** [comment]
@SumanthRH Can you take a look now?

**Review by SumanthRH — COMMENTED** [review]
Did a quick pass, looking good! have you run a successful ppo run with the tinker API server now?

**Comment by SumanthRH** [comment]
To clarify: For the E2E example, 

 You would 
 
 1. Create a new example folder `examples/tinker/ppo`
 
 2. Write a launch script for the tinker api server with the same arguments as in this script which uses our python entrypoint : https://github.com/NovaSky-AI/SkyRL/blob/main/examples/train/ppo/run_ppo.sh
 
 3. Write an example tinker client script for PPO training with SkyRL's Tinker API server in a new `examples/tinker/ppo` folder .

**Comment by tamoghnokandar** [comment]
@SumanthRH Do I have to run the script end-to-end for all epochs? It is taking lot of time to run on 4 GPUs.

**Comment by SumanthRH** [comment]
@tamoghnokandar as long as the first 20 steps matches the training curves as you'd get with the python entrypoint, we should be good

**Comment by tamoghnokandar** [comment]
@SumanthRH This is the curve I am getting. The Tinker curve(in pink) is bit slower than the normal Python entrypoint code (both tested with no KL loss). Maybe due to difference in batching when scheduling requests.Screenshot 2026-04-21 at 2 43 50 AM

**Comment by tamoghnokandar** [comment]
@SumanthRH PPO Critic is still not supported for JAX backend. Maybe this should be done in a separate PR?

**Comment by SumanthRH** [comment]
@tamoghnokandar yes that sounds good to me!

File: tests/backends/test_skyrl_train_backend_loss_validation.py
Comment (SumanthRH) [inline tests/backends/test_skyrl_train_backend_loss_validation.py:1]: @tamoghnokandar this is a pretty hacky test - why are we adding asserts on exact code lines in the script?

Comment (SumanthRH) [reply tests/backends/test_skyrl_train_backend_loss_validation.py:1]: PR looks good otherwise!

Comment (tamoghnokandar) [reply tests/backends/test_skyrl_train_backend_loss_validation.py:1]: Yeah I will remove that test.

**Review by SumanthRH — APPROVED** [review]
Thanks for working on this!

I also did a manual test by running the tinker example script for PPO and comparing the results with the run_ppo.sh script: https://github.com/NovaSky-AI/SkyRL/blob/main/examples/train/ppo/run_ppo.sh

I ran it till 15 steps and metrics including eval reward matched
